[Translation] document `LocaleSwitcher` (#16727)
symfony/symfony-docs: translation.rst

just to run some code. Imagine a console command that renders Twig templates
of emails in different languages. You need to change the locale only to
render those templates.

The ``LocaleSwitcher`` object/service allows you to change at once the locale
of:

1. All the services that are tagged with ``kernel.locale_aware``.
2. ``\Locale::setDefault()``.
3. If a request is available, the ``_locale`` request attribute.

The ``LocaleSwitcher`` service is autowireable and can be injected into other
services::

    use Symfony\Component\Translation\LocaleSwitcher;

    class SomeClass
    {
        public function __construct(
            private LocaleSwitcher $localeSwitcher,
        ) {}

        public function someMethod()
        {
            // you can get the current application locale like this:
            $currentLocale = $this->localeSwitcher->getLocale();

            // you can set the locale for the entire application like this:
            // (from now on, the application will use 'fr' (French) as the
            // locale; including the default locale used to translate Twig templates)
            $this->localeSwitcher->setLocale('fr');

> 

Reviewer: ```suggestion
            // reset the current locale of your application to the configured default locale
            // in config/packages/translation.yaml, by option 'default_locale'
            $this->localeSwitcher->reset();
```
I think it would be interesting to include this function, to indicate that it is possible to reset its locale, by the one configured by default in the application.

What do you think?

Author: Good call, thanks!

Sometimes you need to change the locale of the application dynamically
just to run some code. Imagine a console command that renders Twig templates
of emails in different languages. You need to change the locale only to
render those templates.

The ``LocaleSwitcher`` object/service allows you to change at once the locale
of:

1. All the services that are tagged with ``kernel.locale_aware``.
2. ``\Locale::setDefault()``.
3. If a request is available, the ``_locale`` request attribute.

The ``LocaleSwitcher`` service is autowireable and can be injected into other
services::

    use Symfony\Component\Translation\LocaleSwitcher;

    class SomeClass
    {
        public function __construct(
            private LocaleSwitcher $localeSwitcher,
        ) {}

        public function someMethod()
        {
            // you can get the current application locale like this:
            $currentLocale = $this->localeSwitcher->getLocale();

            // you can set the locale for the entire application like this:
            // (from now on, the application will use 'fr' (French) as the
            // locale; including the default locale used to translate Twig templates)

>             $this->localeSwitcher->setLocale('fr');
>             // reset the current locale of your application to the configured default locale
>             // in config/packages/translation.yaml, by option 'default_locale'
>             $this->localeSwitcher->reset();
>             // you can also run some code with a certain locale, without
>             // changing the locale for the rest of the application
>             $this->localeSwitcher->runWithLocale('es', function() {

Author: ```suggestion
            $this->localeSwitcher->setLocale('fr');

            // reset the current locale of your application to the configured default locale
            // in config/packages/translation.yaml, by option 'default_locale'
            $this->localeSwitcher->reset();

            // you can also run some code with a certain locale, without
            // changing the locale for the rest of the application
            $this->localeSwitcher->runWithLocale('es', function() {
```

    add the missing translation to the log file. For details,
    see :ref:`reference-framework-translator-logging`.

Switch Locale Programmatically
------------------------------

.. versionadded:: 6.1

    The ``LocaleSwitcher`` was introduced in Symfony 6.1.

Sometimes you need to change the locale of the application dynamically
just to run some code. Imagine a console command that renders Twig templates
of emails in different languages. You need to change the locale only to
render those templates.

> The ``LocaleSwitcher`` object/service allows you to change at once the locale

Reviewer 2: ```suggestion
The ``LocaleSwitcher`` class allows you to change at once the locale
```

of:

> 1. All the services that are tagged with ``kernel.locale_aware``.
> 2. ``\Locale::setDefault()``.
> 3. If a request is available, the ``_locale`` request attribute.

Reviewer 2: ```suggestion
* All the services that are tagged with ``kernel.locale_aware``;
* ``\Locale::setDefault()``;
* If a request is available, the ``_locale`` request attribute.
```

The ``LocaleSwitcher`` service is autowireable and can be injected into other

> services::

Reviewer 2: I would remove this bit and add a small paragraph below the example, saying that the service will be autowired when using the default config and noting the service ID to use if that's not the case.

    use Symfony\Component\Translation\LocaleSwitcher;

    class SomeClass
    {
        public function __construct(

>             private LocaleSwitcher $localeSwitcher,

Reviewer 2: I don't think we are using property promotion in the docs already (we usually stick to old practices a little while, as we don't want to distract the reader from the Symfony feature)

    add the missing translation to the log file. For details,
    see :ref:`reference-framework-translator-logging`.

Switch Locale Programmatically
------------------------------

.. versionadded:: 6.1

    The ``LocaleSwitcher`` was introduced in Symfony 6.1.

Sometimes you need to change the locale of the application dynamically
just to run some code. Imagine a console command that renders Twig templates
of emails in different languages. You need to change the locale only to
render those templates.

The ``LocaleSwitcher`` class allows you to change at once the locale
of:

* All the services that are tagged with ``kernel.locale_aware``;
* ``\Locale::setDefault()``;

> * If a request is available, the ``_locale`` request attribute.

Reviewer 3: this is actually wrong. It updates the RequestContext of the Router (so that route generation autofills the `_locale` parameter, making i18n routes work for instance). But the code does not actually updates the Request object.

Author: Thanks, fixed in https://github.com/symfony/symfony-docs/pull/17774.
